SUREFIRE-1250: add exceptions to describe usage rules of %regex

kgyrtkirk opened the pull request:
@Tibor17 i've cleaned it up a bit; and added exception for multiple hashmarks too. tests have passed on my machine.

Tibor17 commented on surefire-api/src/main/java/org/apache/maven/surefire/testset/TestListResolver.java:
```diff
@@ -479,6 +481,27 @@ else if ( hasMethod )
+        {
+            Pattern.compile( classNameExpr );
+            Pattern.compile( methodNameExpr );
+            if ( methodNameExpr.indexOf( "#" ) != -1 )
```
I thought you would use `String.contains` in both constructors of `ResolvedTest`.
I think we do not need to use `Pattern.compile` here because this is already done by `maven-shared-utils` anyway.
Lets use the sanity check if the `isRegex` is set, because I do not know how Cucumber would use JUnit Description of method or class name without regex. Cucumber is different runner from most of the other.

kgyrtkirk commented:
I've used `Pattern.compile` to keep the regexp's problem description; i thinked it makes it more clear how it will be used eventually.

I picked `TestListResolver` as it's place; because this is the first point where the deconstruction logic is used.
plus: I wanted to keep all the generic exceptions of Pattern prior to these construction exceptions - without reassembling the original regex - but maybe this dosen't matter.

I might be able to move it to `ResolvedTest`...tomorrow i will look at it.

kgyrtkirk commented:
i've moved the check into `ResolvedTest`.
I now use the `MatchedPattern`...to check for exceptions.

I've squashed my changes into one commit; because of the sidetrip to `TestListResolver`

Tibor17 commented:
@kgyrtkirk Please close the PR. The code was merged with master. Thx for contributing!

jira-importer commented:
Resolve #2439
